pipeline/transform.py

```diff
@@ -0,0 +1,193 @@
+    """
+    Function to match the string date format to get the 
+    desired datetime format
+    """
+    formats = {
+        "%a, %d %b %Y %H:%M:%S %Z": 
+        r"^[A-Za-z]{3}, \d{2} [A-Za-z]{3} \d{4} \d{2}:\d{2}:\d{2} [A-Z]{3}$",
+        "%Y-%m-%d %H:%M:%S": 
+        r"^\d{4}-\d{2}-\d{2} \d{2}:\d{2}:\d{2}$"
+    }
+    for format_string, pattern in formats.items():
+        if re.match(pattern, date_string):
+            return format_string
+    raise ValueError("Unknown datetime format")
+
+
+def convert_to_datetime(date_string: str) -> None | datetime:
+    """
+    Converts a string into datetime
+    """
+    if not isinstance(date_string, str):
+        return None
+    format_string = identify_datetime_format(date_string)
+    date_time_obj = datetime.strptime(date_string, format_string)
+    return date_time_obj
+
+
+def plant_readings(reading: list[dict]) -> dict:
+    """
+    Returns a dataframe of reading details
+    """
+    if not isinstance(reading, dict):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    return {
+            'soil_moisture': get_details(reading, 'soil_moisture'),
+            'temperature': get_details(reading, 'temperature'),
+            'last_watered': convert_to_datetime(get_details(reading, 'last_watered')),
+            'recording_taken': convert_to_datetime(get_details(reading, 'recording_taken')),
+            'name': get_botanist_detail(reading, 'name'),
+            'plant_name': get_details(reading, 'name')
+        }
+
+
+def group_data(readings: list[dict]) -> tuple[list, list, list]:
+    """
+    Iterates through the readings and organises different parameters
+    into the relevant list
+    """
+    plant = []
+    botanist = []
+    plant_reading = []
+
+    for reading in readings:
+        plant.append(plant_details(reading))
+        botanist.append(botanist_details(reading))
+        plant_reading.append(plant_readings(reading))
+
+    return plant, botanist, plant_reading
+
+
+def convert_to_dataframe(readings: list[dict]) -> pd.DataFrame:
+    """
+    Converts a list into a dataframe
+    """
+    if not isinstance(readings, list):
+        raise TypeError('readings must be of type list')
+    if not all(isinstance(value, dict) for value in readings):
+        raise TypeError('entries into a DataFrame must be of type dict')
+    return pd.DataFrame(readings)
+
+
+def clean_data(df: pd.DataFrame, threshold: int = 2) -> pd.DataFrame:
+    """
+    Drops all 
+    """
+    df = df.dropna(thresh=threshold)
+    return df
+
+
+def transform_data(readings: list[dict]) -> tuple[pd.DataFrame, pd.DataFrame, pd.DataFrame]:
+    """
+    Main
+    """
+    plant, botanist, plant_reading = group_data(readings)
+
+    dim_plant = convert_to_dataframe(plant)
+    dim_botanist = convert_to_dataframe(botanist)
+    fact_plant_reading = convert_to_dataframe(plant_reading)
+
+    return clean_data(dim_plant), clean_data(dim_botanist), clean_data(fact_plant_reading)
```

Comment:
Adding type checking in `group_data` will mean you don't have to type check here